src/lib/db/feature-toggle-store.ts

```diff
@@ -171,34 +173,40 @@ export default class FeatureToggleStore implements IFeatureToggleStore {
 
     async setLastSeen(data: LastSeenInput[]): Promise<void> {
         const now = new Date();
-        const environmentArrays = this.mapMetricDataToEnvBuckets(data);
-        try {
-            for (const env of Object.keys(environmentArrays)) {
-                const toggleNames = environmentArrays[env].sort();
-                await this.db(FEATURE_ENVIRONMENTS_TABLE)
-                    .update({ last_seen_at: now })
-                    .where('environment', env)
-                    .whereIn(
-                        'feature_name',
-                        this.db(FEATURE_ENVIRONMENTS_TABLE)
-                            .select('feature_name')
-                            .whereIn('feature_name', toggleNames)
-                            .forUpdate()
-                            .skipLocked(),
-                    );
+        const enhancedData = data.map((value) => ({
+            feature_name: value.featureName,
+            last_seen_at: now,
+            environment: value.environment,
+        }));
 
-                // Updating the toggle's last_seen_at also for backwards compatibility
-                await this.db(TABLE)
-                    .update({ last_seen_at: now })
+        try {
+            await this.db.transaction(async (trx) => {
+                // Create a CTE with the data to be inserted
+                const cte = trx('features')
+                    .select('name')
                     .whereIn(
                         'name',
-                        this.db(TABLE)
-                            .select('name')
-                            .whereIn('name', toggleNames)
-                            .forUpdate()
-                            .skipLocked(),
+                        enhancedData.map((entry) => entry.feature_name),
                     );
-            }
+
+                // Insert only the rows that exist in the CTE (subquery)
```

Comment:
If the features table is only used in the CTE subquery, the likelihood of deadlocks occurring directly due to this table is significantly reduced.